File: solutions/ruby/protein-translation/1/protein_translation.rb
class InvalidCodonError < StandardError
end

class Translation
  def self.of_rna(strand)
    res = []
    
    (0...strand.length).step(3).each do |i|
      codon = strand[i...i+3]
      p = protien(codon)
      
      break if p === 'STOP'
      
      res << protien(codon)
    end

    res
  end

  private_class_method def self.protien(codon)
    case codon
      when 'AUG' then 'Methionine'
      when 'UUU', 'UUC' then 'Phenylalanine'
      when 'UUA', 'UUG' then 'Leucine'
      when 'UCU', 'UCC', 'UCA', 'UCG' then 'Serine'
      when 'UAU', 'UAC' then 'Tyrosine'
      when 'UGU', 'UGC' then 'Cysteine'
      when 'UGG' then 'Tryptophan'
      when 'UAA', 'UAG', 'UGA' then 'STOP'
      else raise InvalidCodonError
    end
  end
end

File: solutions/go/protein-translation/1/protein_translation.go
package protein

import "errors"

var ErrStop error = errors.New("Stop sequence encountered")
var ErrInvalidBase error = errors.New("InvalidBase")

func FromRNA(rna string) ([]string, error) {
	codonCount := len(rna) / 3
	res := []string{}
	var codon string
	var err error
	var startIndex int
	var endIndex int

	for i:=0; i < codonCount; i++ {
		startIndex =  i * 3
		endIndex = (i + 1) * 3

		codon, err = FromCodon(rna[startIndex:endIndex])

		if err == ErrStop {
			return res, nil
		} else if err == ErrInvalidBase {
			return []string{}, err
		}

		res = append(res, codon)
	}

	return res, nil
}

func FromCodon(codon string) (string, error) {
	switch {
	case codon == "AUG":
		return "Methionine", nil
	case codon == "UUU" || codon == "UUC":
		return "Phenylalanine", nil
	case codon == "UUA" || codon == "UUG":
		return "Leucine", nil
	case codon == "UCU" || codon == "UCC" || codon == "UCA" || codon == "UCG":
		return "Serine", nil
	case codon == "UAU" || codon == "UAC":
		return "Tyrosine", nil
	case codon == "UGU" || codon == "UGC":
		return "Cysteine", nil
	case codon == "UGG":
		return "Tryptophan", nil
	case codon == "UGU" || codon == "UGC":
		return "Cysteine", nil
	case codon == "UAA" || codon == "UAG" || codon == "UGA":
		return "", ErrStop
	default:
		return "", ErrInvalidBase
	}
}
